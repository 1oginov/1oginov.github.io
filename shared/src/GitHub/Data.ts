import { Repo } from './Repo';

export interface Data {
  homepageUrl: string;
  login: string;
  repos: Repo[];
  timestamp: number;
  url: string;
}
