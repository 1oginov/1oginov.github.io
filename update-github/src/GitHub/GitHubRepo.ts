/* eslint-disable camelcase */

export interface GitHubRepo {
  archived: boolean;
  description: string | null;
  html_url: string;
  homepage: string | null;
  language: string | null;
  name: string;
  owner: {
    login: string;
  };
  stargazers_count: number;
  tags_url: string;
  updated_at: string;
}
