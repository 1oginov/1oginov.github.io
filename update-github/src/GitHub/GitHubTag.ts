/* eslint-disable camelcase */

export interface GitHubTag {
  name: string;
}
