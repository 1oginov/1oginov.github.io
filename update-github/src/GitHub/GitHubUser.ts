/* eslint-disable camelcase */

export interface GitHubUser {
  blog: string;
  html_url: string;
  login: string;
  repos_url: string;
}
