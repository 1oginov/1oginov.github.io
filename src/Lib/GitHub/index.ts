import { USERNAMES } from 'Constants';

import GitHub from './GitHub';

const instance = new GitHub({ username: USERNAMES.GITHUB });

export default instance;
