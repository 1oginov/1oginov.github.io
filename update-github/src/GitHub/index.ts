import { GITHUB_PERSONAL_ACCESS_TOKEN } from 'Constants';

import { GitHub } from './GitHub';

const instance = new GitHub({ personalAccessToken: GITHUB_PERSONAL_ACCESS_TOKEN });

export default instance;
