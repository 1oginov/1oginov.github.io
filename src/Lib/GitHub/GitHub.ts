import GitHubOptions from './GitHubOptions';
import GitHubRepo from './GitHubRepo';

export default class GitHub {
  private readonly baseUrl: string = 'https://api.github.com';

  private readonly username: string;

  private static parseNextPageUrl(headers: Headers): string | undefined {
    const header = headers.get('Link');

    if (!header) {
      return undefined;
    }

    const links = header.split(',');
    const link = links.find((l) => l.endsWith('>; rel="next"'));

    if (!link) {
      return undefined;
    }

    return link.trim().substring(1).replace('>; rel="next"', '');
  }

  private static fetchAll(url: string, responsesStack: Response[] = []): Promise<Response[]> {
    return fetch(url).then((response) => {
      responsesStack.push(response);

      const nextPage = this.parseNextPageUrl(response.headers);

      if (nextPage) {
        return this.fetchAll(nextPage, responsesStack);
      }

      return responsesStack;
    });
  }

  private static mergeJsonResponses<T>(responsesStack: Response[]): Promise<T> {
    return Promise.all(responsesStack.map((response) => response.json()))
      .then((results) => {
        const [firstResult, ...otherResults] = results;

        if (otherResults) {
          return firstResult.concat(...otherResults);
        }

        return firstResult;
      });
  }

  constructor({ username }: GitHubOptions) {
    this.username = username;
  }

  getRepos(): Promise<GitHubRepo[]> {
    // Avoid GitHub API throttling while developing.
    if (process.env.NODE_ENV === 'development') {
      // eslint-disable-next-line @typescript-eslint/no-var-requires, global-require
      return Promise.resolve(require('./__fixtures__/repos.json'));
    }

    return GitHub.fetchAll(`${this.baseUrl}/users/${this.username}/repos`)
      .then((responses) => GitHub.mergeJsonResponses<GitHubRepo[]>(responses))
      .catch(() => []);
  }
}
