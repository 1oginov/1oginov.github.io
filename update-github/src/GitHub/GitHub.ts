import { GitHubData as Data, GitHubRepo as Repo } from '@loginov-rocks/loginov-rocks-shared';
import fetch, { Headers, Response } from 'node-fetch';

import { GitHubRepo } from './GitHubRepo';
import { GitHubUser } from './GitHubUser';

interface Options {
  baseUrl: string;
  personalAccessToken: string;
}

interface GetReposFilter {
  // eslint-disable-next-line no-unused-vars
  (gitHubRepo: GitHubRepo): boolean;
}

export class GitHub {
  private static parseNextPageUrl(headers: Headers): string | undefined {
    const header = headers.get('Link');

    if (!header) {
      return undefined;
    }

    const links = header.split(',');
    const link = links.find((l) => l.endsWith('>; rel="next"'));

    if (!link) {
      return undefined;
    }

    return link.trim().substring(1).replace('>; rel="next"', '');
  }

  private static async mergeArrayResponses<T>(responsesStack: Response[]): Promise<T> {
    const results = await Promise.all(responsesStack.map((response) => response.json()));

    const [firstResult, ...otherResults] = results;

    if (otherResults) {
      return firstResult.concat(...otherResults);
    }

    return firstResult;
  }

  private readonly baseUrl: string;

  private readonly personalAccessToken: string;

  constructor({ baseUrl, personalAccessToken }: Options) {
    this.baseUrl = baseUrl;
    this.personalAccessToken = personalAccessToken;
  }

  private apiGet(url: string): Promise<Response> {
    const apiUrl = url.startsWith(this.baseUrl) ? url : `${this.baseUrl}${url}`;

    return fetch(apiUrl, {
      headers: {
        Authorization: `token ${this.personalAccessToken}`,
      },
    });
  }

  private async apiGetWithPagination(url: string, responsesStack: Response[] = []): Promise<Response[]> {
    const response = await this.apiGet(url);

    responsesStack.push(response);

    const nextPageUrl = GitHub.parseNextPageUrl(response.headers);

    if (nextPageUrl) {
      return this.apiGetWithPagination(nextPageUrl, responsesStack);
    }

    return responsesStack;
  }

  private async getGitHubUser(): Promise<GitHubUser> {
    const response = await this.apiGet('/user');

    return response.json();
  }

  private async getGitHubRepos(): Promise<GitHubRepo[]> {
    const responsesStack = await this.apiGetWithPagination('/user/repos');

    return GitHub.mergeArrayResponses<GitHubRepo[]>(responsesStack);
  }

  private async getRepos(filter: GetReposFilter = () => true): Promise<Repo[]> {
    const gitHubRepos = await this.getGitHubRepos();

    return gitHubRepos
      .filter(filter)
      .map((repo) => ({
        description: repo.description || '',
        homepageUrl: repo.homepage || '',
        isArchived: repo.archived,
        language: repo.language || '',
        // TODO
        latestTag: '',
        name: repo.name,
        stars: repo.stargazers_count,
        updatedAt: new Date(repo.updated_at).getTime(),
        url: repo.html_url,
      }));
  }

  async getData(): Promise<Data> {
    const gitHubUser = await this.getGitHubUser();
    const repos = await this.getRepos((gitHubRepo) => gitHubRepo.owner.login === gitHubUser.login);

    return {
      homepageUrl: gitHubUser.blog,
      login: gitHubUser.login,
      repos,
      timestamp: Date.now(),
      url: gitHubUser.html_url,
    };
  }
}
