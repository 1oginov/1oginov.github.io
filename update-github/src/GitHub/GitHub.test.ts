import { GitHub } from './GitHub';

jest.mock('@octokit/rest', () => ({
  Octokit: class {
    // eslint-disable-next-line class-methods-use-this
    request(url: string): Promise<unknown> {
      if (url === '/user') {
        // eslint-disable-next-line @typescript-eslint/no-var-requires, global-require
        const data = require('./__fixtures__/gitHubUser.json');
        return Promise.resolve({ data });
      }

      if (url === 'https://api.github.com/users/loginov-rocks/repos') {
        // eslint-disable-next-line @typescript-eslint/no-var-requires, global-require
        const data = require('./__fixtures__/gitHubRepos.json');
        return Promise.resolve({ data });
      }

      return Promise.reject();
    }
  },
}));

const gitHub = new GitHub({ personalAccessToken: 'personal-access-token' });

describe('getData', () => {
  it('gets GitHub data using API', async () => {
    const data = await gitHub.getData();

    expect(data).toStrictEqual(
      expect.objectContaining({
        homepageUrl: 'https://loginov.rocks',
        url: 'https://github.com/loginov-rocks',
        user: 'loginov-rocks',
      }),
    );
    expect(typeof data.timestamp).toBe('number');

    expect(Array.isArray(data.repos)).toBeTruthy();
    expect(data.repos.length).toBe(30);

    expect(data.repos[0]).toStrictEqual({
      description: 'Clean but full featured AngularJS boilerplate using Gulp workflow and best practices',
      homepageUrl: 'https://www.npmjs.com/package/angular-gulp-boilerplate',
      language: 'JavaScript',
      latestVersion: '',
      stars: 29,
      title: 'Angular-Gulp-Boilerplate',
      updatedAt: 1613486363000,
      url: 'https://github.com/loginov-rocks/Angular-Gulp-Boilerplate',
    });

    expect(data.repos).toMatchSnapshot();
  });
});
