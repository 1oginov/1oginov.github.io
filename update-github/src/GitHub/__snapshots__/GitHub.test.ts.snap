// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getData gets GitHub data using API 1`] = `
Array [
  Object {
    "description": "Clean but full featured AngularJS boilerplate using Gulp workflow and best practices",
    "homepageUrl": "https://www.npmjs.com/package/angular-gulp-boilerplate",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 29,
    "title": "Angular-Gulp-Boilerplate",
    "updatedAt": 1613486363000,
    "url": "https://github.com/loginov-rocks/Angular-Gulp-Boilerplate",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Anticloud",
    "updatedAt": 1610848376000,
    "url": "https://github.com/loginov-rocks/Anticloud",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "anticloud-server",
    "updatedAt": 1606655910000,
    "url": "https://github.com/loginov-rocks/anticloud-server",
  },
  Object {
    "description": "ES6 class for serial communication with your own Bluetooth Low Energy (Smart) devices",
    "homepageUrl": "https://loginov-rocks.github.io/Web-Bluetooth-Terminal/",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 31,
    "title": "bluetooth-terminal",
    "updatedAt": 1610847914000,
    "url": "https://github.com/loginov-rocks/bluetooth-terminal",
  },
  Object {
    "description": "Canvas heightmap generator",
    "homepageUrl": "https://www.npmjs.com/package/canvas-heightmap",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 1,
    "title": "canvas-heightmap",
    "updatedAt": 1609888189000,
    "url": "https://github.com/loginov-rocks/canvas-heightmap",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Carbook",
    "updatedAt": 1610848388000,
    "url": "https://github.com/loginov-rocks/Carbook",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "Dart",
    "latestVersion": "",
    "stars": 0,
    "title": "carquery",
    "updatedAt": 1610848367000,
    "url": "https://github.com/loginov-rocks/carquery",
  },
  Object {
    "description": "Extract comments from glob of files, annotated or not",
    "homepageUrl": "https://www.npmjs.com/package/comments-extractor",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "comments-extractor",
    "updatedAt": 1609802038000,
    "url": "https://github.com/loginov-rocks/comments-extractor",
  },
  Object {
    "description": "",
    "homepageUrl": "",
    "isArchived": false,
    "language": "C++",
    "latestVersion": "",
    "stars": 0,
    "title": "Connected-Espresso-Machine",
    "updatedAt": 1609070113000,
    "url": "https://github.com/loginov-rocks/Connected-Espresso-Machine",
  },
  Object {
    "description": "Simple app for communication with devices via bluetooth based on Apache Cordova (PhoneGap)",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 10,
    "title": "Cordova-Bluetooth-Terminal",
    "updatedAt": 1568776082000,
    "url": "https://github.com/loginov-rocks/Cordova-Bluetooth-Terminal",
  },
  Object {
    "description": "ðŸ¦‹ An OpenApi validator for ExpressJS that automatically validates API requests using an OpenAPI 3.x specification,",
    "homepageUrl": "",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "express-openapi-validator",
    "updatedAt": 1567085587000,
    "url": "https://github.com/loginov-rocks/express-openapi-validator",
  },
  Object {
    "description": "Firebase Authenticator",
    "homepageUrl": "https://loginov.rocks/Firebase-Authenticator/",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Firebase-Authenticator",
    "updatedAt": 1606655254000,
    "url": "https://github.com/loginov-rocks/Firebase-Authenticator",
  },
  Object {
    "description": "Gulp plugin to bundle JSON localization files",
    "homepageUrl": "https://www.npmjs.com/package/gulp-locales-bundler",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 1,
    "title": "gulp-locales-bundler",
    "updatedAt": 1609889891000,
    "url": "https://github.com/loginov-rocks/gulp-locales-bundler",
  },
  Object {
    "description": "Documentation generator for Gulp tasks",
    "homepageUrl": "https://www.npmjs.com/package/gulpdoc",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "gulpdoc",
    "updatedAt": 1609802952000,
    "url": "https://github.com/loginov-rocks/gulpdoc",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Igla",
    "updatedAt": 1610848398000,
    "url": "https://github.com/loginov-rocks/Igla",
  },
  Object {
    "description": "",
    "homepageUrl": "",
    "isArchived": false,
    "language": "C++",
    "latestVersion": "",
    "stars": 0,
    "title": "IoT-Firebase-Authentication",
    "updatedAt": 1574692842000,
    "url": "https://github.com/loginov-rocks/IoT-Firebase-Authentication",
  },
  Object {
    "description": "Locales bundler",
    "homepageUrl": "https://www.npmjs.com/package/locales-bundler",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "locales-bundler",
    "updatedAt": 1609890326000,
    "url": "https://github.com/loginov-rocks/locales-bundler",
  },
  Object {
    "description": "MeArm Controller",
    "homepageUrl": "",
    "isArchived": false,
    "language": "Arduino",
    "latestVersion": "",
    "stars": 2,
    "title": "MeArm-Controller",
    "updatedAt": 1569915054000,
    "url": "https://github.com/loginov-rocks/MeArm-Controller",
  },
  Object {
    "description": "MQTT Home Monitor",
    "homepageUrl": "",
    "isArchived": false,
    "language": "C++",
    "latestVersion": "",
    "stars": 0,
    "title": "MQTT-Home-Monitor",
    "updatedAt": 1587220046000,
    "url": "https://github.com/loginov-rocks/MQTT-Home-Monitor",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Node-AWS-App",
    "updatedAt": 1610848267000,
    "url": "https://github.com/loginov-rocks/Node-AWS-App",
  },
  Object {
    "description": "Quick start with NodeMCU v3 (ESP8266), Arduino ecosystem, and PlatformIO IDE",
    "homepageUrl": "https://loginov-rocks.medium.com/quick-start-with-nodemcu-v3-esp8266-arduino-ecosystem-and-platformio-ide-b8415bf9a038",
    "isArchived": false,
    "language": "C++",
    "latestVersion": "",
    "stars": 2,
    "title": "NodeMCU-Arduino-PlatformIO",
    "updatedAt": 1608951885000,
    "url": "https://github.com/loginov-rocks/NodeMCU-Arduino-PlatformIO",
  },
  Object {
    "description": "Ooyala Player Component",
    "homepageUrl": "https://www.npmjs.com/package/ooyala-player-component",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "ooyala-player-component",
    "updatedAt": 1609892621000,
    "url": "https://github.com/loginov-rocks/ooyala-player-component",
  },
  Object {
    "description": "Portfolio App",
    "homepageUrl": "https://loginov-rocks.github.io/Portfolio/",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Portfolio",
    "updatedAt": 1610903025000,
    "url": "https://github.com/loginov-rocks/Portfolio",
  },
  Object {
    "description": "Firebase Services for Portfolio App",
    "homepageUrl": "https://loginov.rocks/Portfolio",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "Portfolio-Firebase",
    "updatedAt": 1606654675000,
    "url": "https://github.com/loginov-rocks/Portfolio-Firebase",
  },
  Object {
    "description": "React Native Redux Form fields",
    "homepageUrl": "https://www.npmjs.com/package/react-native-redux-form",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 0,
    "title": "react-native-redux-form",
    "updatedAt": 1609890994000,
    "url": "https://github.com/loginov-rocks/react-native-redux-form",
  },
  Object {
    "description": "Redux Repository",
    "homepageUrl": "https://www.npmjs.com/package/redux-repository",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "redux-repository",
    "updatedAt": 1609893885000,
    "url": "https://github.com/loginov-rocks/redux-repository",
  },
  Object {
    "description": "Three.js Sky object from official repository example",
    "homepageUrl": "https://www.npmjs.com/package/three-sky",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 2,
    "title": "three-sky",
    "updatedAt": 1609888944000,
    "url": "https://github.com/loginov-rocks/three-sky",
  },
  Object {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "TypeScript",
    "latestVersion": "",
    "stars": 0,
    "title": "todogle",
    "updatedAt": 1610848310000,
    "url": "https://github.com/loginov-rocks/todogle",
  },
  Object {
    "description": "Arduino library for the fastest and simplest communication with u-blox GPS modules",
    "homepageUrl": "",
    "isArchived": false,
    "language": "C++",
    "latestVersion": "",
    "stars": 83,
    "title": "UbxGps",
    "updatedAt": 1613437756000,
    "url": "https://github.com/loginov-rocks/UbxGps",
  },
  Object {
    "description": "Progressive Web Application for serial communication with your own Bluetooth Low Energy (Smart) devices",
    "homepageUrl": "https://loginov-rocks.github.io/Web-Bluetooth-Terminal/",
    "isArchived": false,
    "language": "JavaScript",
    "latestVersion": "",
    "stars": 129,
    "title": "Web-Bluetooth-Terminal",
    "updatedAt": 1613322210000,
    "url": "https://github.com/loginov-rocks/Web-Bluetooth-Terminal",
  },
]
`;
